[mqtt.homeassistant] discovery not working when device object is set (correctly)

The homeassistant autodiscovery process expects a json payload that describes the device. 

The correct format for the device part of the message should look like this:
```json
"device":{"identifiers":["1A32E3"],"connections":[["mac","5C:CF:7F:1A:32:E3"]]}
``` 
Notice the double [ brackets. It is a list in a list (specified as a tuple in a list).  This is correct according to the specifications here:

https://www.home-assistant.io/integrations/sensor.mqtt/#connections

However when a message includes this part and is formatted like that, the binding will fail with a message like this:
<details>
<summary>click to open log (hidden because it is long)...</summary>
<p>

```log
2019-12-11 10:57:13.026 [ERROR] [ansport.mqtt.internal.ClientCallback] - MQTT message received. MqttMessageSubscriber#processMessage() implementation failure
com.google.gson.JsonSyntaxException: java.lang.IllegalStateException: Expected BEGIN_OBJECT but was BEGIN_ARRAY at line 1 column 324 path $.device.connections[0]
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:224) ~[?:?]
	at com.google.gson.internal.bind.TypeAdapterRuntimeTypeWrapper.read(TypeAdapterRuntimeTypeWrapper.java:41) ~[?:?]
	at com.google.gson.internal.bind.CollectionTypeAdapterFactory$Adapter.read(CollectionTypeAdapterFactory.java:82) ~[?:?]
	at com.google.gson.internal.bind.CollectionTypeAdapterFactory$Adapter.read(CollectionTypeAdapterFactory.java:61) ~[?:?]
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.read(ReflectiveTypeAdapterFactory.java:129) ~[?:?]
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:220) ~[?:?]
	at org.openhab.binding.mqtt.homeassistant.internal.ChannelConfigurationTypeAdapterFactory$2.read(ChannelConfigurationTypeAdapterFactory.java:101) ~[?:?]
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.read(ReflectiveTypeAdapterFactory.java:129) ~[?:?]
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:220) ~[?:?]
	at org.openhab.binding.mqtt.homeassistant.internal.ChannelConfigurationTypeAdapterFactory$1.read(ChannelConfigurationTypeAdapterFactory.java:77) ~[?:?]
	at com.google.gson.Gson.fromJson(Gson.java:888) ~[?:?]
	at com.google.gson.Gson.fromJson(Gson.java:853) ~[?:?]
	at com.google.gson.Gson.fromJson(Gson.java:802) ~[?:?]
	at com.google.gson.Gson.fromJson(Gson.java:774) ~[?:?]
	at org.openhab.binding.mqtt.homeassistant.internal.BaseChannelConfiguration.fromString(BaseChannelConfiguration.java:57) ~[?:?]
	at org.openhab.binding.mqtt.homeassistant.internal.BaseChannelConfiguration.fromString(BaseChannelConfiguration.java:68) ~[?:?]
	at org.openhab.binding.mqtt.homeassistant.internal.discovery.HomeAssistantDiscovery.receivedMessage(HomeAssistantDiscovery.java:149) ~[?:?]
	at org.openhab.binding.mqtt.discovery.TopicSubscribe.processMessage(TopicSubscribe.java:53) ~[?:?]
	at org.eclipse.smarthome.io.transport.mqtt.internal.ClientCallback.lambda$3(ClientCallback.java:96) ~[?:?]
	at java.util.ArrayList.forEach(ArrayList.java:1257) ~[?:1.8.0_222]
	at org.eclipse.smarthome.io.transport.mqtt.internal.ClientCallback.messageArrived(ClientCallback.java:96) ~[?:?]
	at org.eclipse.smarthome.io.transport.mqtt.internal.ClientCallback.messageArrived(ClientCallback.java:71) ~[?:?]
	at com.hivemq.client.internal.mqtt.mqtt3.Mqtt3AsyncClientView.lambda$callbackView$1(Mqtt3AsyncClientView.java:73) ~[bundleFile:?]
	at com.hivemq.client.internal.mqtt.MqttAsyncClient$CallbackSubscriber.onNext(MqttAsyncClient.java:227) [bundleFile:?]
	at com.hivemq.client.internal.mqtt.MqttAsyncClient$CallbackSubscriber.onNext(MqttAsyncClient.java:212) [bundleFile:?]
	at com.hivemq.client.rx.FlowableWithSingle$SingleFutureSubscriber.onNext(FlowableWithSingle.java:377) [bundleFile:?]
	at com.hivemq.client.internal.rx.operators.FlowableWithSingleCombine$SplitSubscriber$Default.tryOnNextActual(FlowableWithSingleCombine.java:206) [bundleFile:?]
	at com.hivemq.client.internal.rx.operators.FlowableWithSingleCombine$SplitSubscriber.tryOnNext(FlowableWithSingleCombine.java:171) [bundleFile:?]
	at io.reactivex.internal.operators.flowable.FlowableObserveOn$ObserveOnConditionalSubscriber.runAsync(FlowableObserveOn.java:649) [bundleFile:?]
	at io.reactivex.internal.operators.flowable.FlowableObserveOn$BaseObserveOnSubscriber.run(FlowableObserveOn.java:176) [bundleFile:?]
	at io.reactivex.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:66) [bundleFile:?]
	at io.reactivex.internal.schedulers.ScheduledRunnable.call(ScheduledRunnable.java:57) [bundleFile:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_222]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) [?:1.8.0_222]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) [?:1.8.0_222]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_222]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_222]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_222]
Caused by: java.lang.IllegalStateException: Expected BEGIN_OBJECT but was BEGIN_ARRAY at line 1 column 324 path $.device.connections[0]
	at com.google.gson.stream.JsonReader.beginObject(JsonReader.java:385) ~[?:?]
	at org.openhab.binding.mqtt.generic.tools.JsonReaderDelegate.beginObject(JsonReaderDelegate.java:63) ~[?:?]
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:213) ~[?:?]
	... 37 more
```

</p>
</details> 

This can see be seen that the expected json is
```json
"device": {
"connections": [ {"type":"mac", "identifier" : "00:00:00:00"} ]
}
```
from 
https://github.com/openhab/openhab2-addons/blob/746ecb7a8f856f8e21ca31b61b5edd4c6a65b5b4/bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/BaseChannelConfiguration.java#L96-L114

--- a/bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/BaseChannelConfiguration.java
+++ b/bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/BaseChannelConfiguration.java
@@ -108,6 +108,7 @@ public abstract class BaseChannelConfiguration {
         }
     }
 
+    @JsonAdapter(ConnectionDeserializer.class)
     static class Connection {
         protected @Nullable String type;
         protected @Nullable String identifier;

--- a/bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/ConnectionDeserializer.java
+++ b/bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/ConnectionDeserializer.java
@@ -1,0 +1,36 @@
+/**
+ * Copyright (c) 2010-2019 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.mqtt.homeassistant.internal;
+
+import com.google.gson.*;
+
+import java.lang.reflect.Type;
+
+/**
+ * The {@link ConnectionDeserializer} will de-serialize a connection-list
+ *
+ * see: https://www.home-assistant.io/integrations/sensor.mqtt/#connections
+ *
+ * @author Jan N. Klug - Initial contribution
+ */
+public class ConnectionDeserializer implements JsonDeserializer<BaseChannelConfiguration.Connection> {
+    @Override
+    public BaseChannelConfiguration.Connection deserialize(JsonElement json, Type typeOfT,
+            JsonDeserializationContext context) throws JsonParseException {
+        JsonArray list = json.getAsJsonArray();
+        BaseChannelConfiguration.Connection conn = new BaseChannelConfiguration.Connection();
+        conn.type = list.get(0).getAsString();
+        conn.identifier = list.get(1).getAsString();
+        return conn;
+    }
+}

--- a/bundles/org.openhab.binding.mqtt.homeassistant/src/test/resources/org/openhab/binding/mqtt/homeassistant/internal/configA.json
+++ b/bundles/org.openhab.binding.mqtt.homeassistant/src/test/resources/org/openhab/binding/mqtt/homeassistant/internal/configA.json
@@ -13,10 +13,10 @@
 			"H"
 		],
 		"cns": [
-			{
-				"type": "I1",
-				"identifier": "I2"
-			}
+			[
+				"I1",
+				"I2"
+			]
 		],
 		"name": "J",
 		"mdl": "K",

--- a/bundles/org.openhab.binding.mqtt.homeassistant/src/test/resources/org/openhab/binding/mqtt/homeassistant/internal/configB.json
+++ b/bundles/org.openhab.binding.mqtt.homeassistant/src/test/resources/org/openhab/binding/mqtt/homeassistant/internal/configB.json
@@ -14,10 +14,10 @@
 	"device": {
 		"ids": "H",
 		"cns": [
-			{
-				"type": "I1",
-				"identifier": "I2"
-			}
+			[
+				"I1",
+				"I2"
+			]
 		],
 		"name": "J",
 		"mdl": "K",

